src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,287 @@
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
+#include "zcash/IncrementalMerkleTree.hpp"
+#include "crypto/sha256.h"
+#include "zerocash/utils/util.h" // TODO: remove these utilities
+
+namespace libzcash {
+
+SHA256Compress SHA256Compress::combine(const SHA256Compress& a, const SHA256Compress& b)
+{
+    // This is a performance optimization.
+    if (a.IsNull() && b.IsNull()) {
+        return a;
+    }
+
+    SHA256Compress res = SHA256Compress();
+
+    CSHA256 hasher;
+    hasher.Write(a.begin(), 32);
+    hasher.Write(b.begin(), 32);
+    hasher.FinalizeNoPadding(res.begin());
+
+    return res;
+}
+
+
+template <typename Hash>
+class PathFiller {
+private:
+    std::deque<Hash> queue;
+public:
+    PathFiller() : queue() { }
+    PathFiller(std::deque<Hash> queue) : queue(queue) { }
+
+    Hash next() {
+        if (queue.size() > 0) {
+            Hash h = queue.front();
+            queue.pop_front();
+
+            return h;
+        } else {
+            return Hash();
+        }
+    }
+};
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::wfcheck() const {
+    if (parents.size() >= Depth) {
+        throw std::ios_base::failure("tree has too many parents");
+    }
+}
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::append(Hash obj) {
+    if (is_complete(Depth)) {
+        throw std::runtime_error("tree is full");
+    }
+
+    if (!left) {
+        // Set the left leaf
+        left = obj;
+    } else if (!right) {
+        // Set the right leaf
+        right = obj;
+    } else {
+        // Combine the leaves and propagate it up the tree
+        boost::optional<Hash> combined = Hash::combine(*left, *right);
+
+        // Set the left leaf to the object and make the right object none
+        left = obj;
+        right = boost::none;
+
+        // Propagate up the tree as much as needed
+        BOOST_FOREACH(boost::optional<Hash>& parent, parents) {
+            if (parent) {
+                combined = Hash::combine(*parent, *combined);
+                parent = boost::none;
+            } else {
+                parent = *combined;
+                combined = boost::none;
+                break;
+            }
+        }
+
+        if (combined) {
+            // Create a new parent
+            parents.push_back(combined);
+        }
+    }
+}
```

Comment:
It took me a while to understand this, but I think I get it now:

Except for the first element, adding an element always completes the rightmost depth-1 subtree. Completing that subtree *may* complete the depth-2 subtree containing it, and completing *that* subtree *may* complete the depth-3 subtree containing *it*.

It's easier to imagine parents starting out as an array of none's.

parents[0] is always the root of the rightmost depth-1 tree. If it's none, then that means we're on the *left* depth-1 tree in the rightmost depth-2 tree ("none" meaning "we don't know the root yet"). If it's set, then we're on the *right* depth-1 tree in the rightmost depth-2 tree and it contains the root of the depth-1 tree to our left.

To understand what parents[0] means, imagine the rightmost depth-2 subtree. That subtree has two depth-1 subtree children, one on the left, one on the right. parents[0] is the root of the one on the left. When it's none, it means "we don't know the value of that root yet" because we haven't filled in its left/right children. When it's not-none, it's the root of that subtree.

parents[1] is similarly the root (either known or unknown) of the leftmost depth-2 subtree within the rightmost depth-3 subtree.

The `BOOST_FOREACH` loop is like this: We start out having just completed a depth-1 subtree. If `parent[0]` is set, then we're a right-hand depth-1 subtree so we also just completed a depth-2 subtree so save its root and continue. Continuing, we check if we're a right-hand depth-2 subtree, if we are we just completed a depth-3 subtree so continue. And so on... If we ever encounter a none, it means we didn't complete a subtree so we just save the root of the one we completed.

I wonder if there's a way to make how this works more clear.